Plot update

lghszdnp opened the pull request.

lghszdnp committed:
Plot update

neilhatfield committed:
Convert to app.R

This is the conversion to the app.R framework; you will need to address requested changes from this point.

neilhatfield reviewed (changes requested):
There are several changes that I'm requesting you make at this time. However, be sure that you pull the most recent version of the code I pushed to GitHub (i.e., the app.R framework; commit a1c368ca57f5d54d7e5443769d3ee95640fb3165).

1. Be sure to update your R and all packages to the latest versions.
2. Be sure to set `APP_TITLE`
3. Set `id = "pages"` for the `sidebarmenu` and make all necessary changes to buttons that deal with sending the user somewhere (e.g., the go button)
4. Add the comment bubble/app feedback to the header.
5. Buttons
    a. The "Explore" button on the first page makes a user anticipate that they are going to the Explore tab, but instead takes them to Prereq's. I recommend that you either change the label if you want to send the user to Prereq's OR change where that button sends users.
    b. Given the many places a person can go, what would you think about having more buttons on the Overview page OR no buttons but a reference to the sidebar?
    c. I think a button on the Prereq's page going to the Explore page would be useful.
    d. Use `size = "large"` for all bsButtons as a default
    e. For the submit button, use `style = "default"` instead of "warning"
6. Please complete spelling and grammar checks; I noticed the following errors; there might be more
    a. "Assumptiopns" in the title
    b. "InValid" for the ANOVA tab of the Explore page
    c. Sentences missing end of sentence punctuation.
    d. There are many other inconsistencies, spelling/grammar problems throughout much of the app.
7. You don't need "We would like to extend a...." in your Acknowledgements and you don't really need to acknowledge Bob or I unless we did something that fundamentally helped you.
8. References
    a. Update the reference list to include what packages you used
    b. Put all references in AmStat style (see [Style Guide-References](https://educationshinyappteam.github.io/Style_Guide/documentation.html))
    c. Remove the reference to the Caveats app
    d. You'll need to add the following:
      - Hatfield, N. J. (2020), "Stat 461: ANOVA Course Notes" [course notes], Spring 2020.
      - Oehlert, G. W. (2000), "A First Course in Design and Analysis of Experiments" [keyboarding data set], New York: W. H. Freeman.
      - Kutner, M. H., Nachtsheim, C. J., Neter, J., and Li, W. (2005), "Applied Linear Statistical Models" [apex enterprises data set], New York: McGraw-Hill Irwin
9. Code formatting
    a. Watch out for long code lines (i.e. over 80 characters). Turn on RStudio's guide to help you find the lines. Keep in mind that there are some things that you break, for example, URLs; these will will have to be long.
    b. Be sure to be explicit with arguments. For example, "6" listed without the argument inside of `column` or `updateButton` with no arguments named.
    b. Generally speaking, unless all arguments fit on the same line, you can make your code more readable by giving each argument its own line with the first argument being listed below the function name. (Note:  you may have the `dashboardHeader` and `dashboardSidebar` elements violate this; however, the `dashboardBody` is where this matters). For example,

```
div(
  style = "text-align: center;",
  bsButton(
    inputId = "explore",
    etc.
```

10. HTML & CSS Coding
    a. Review [Style Guide-HTML](https://educationshinyappteam.github.io/Style_Guide/coding.html#html). You can't mix tags together such as `p(tags$li("One field has a higher clay content...`. 
    b. You should order tags in terms of their scope; i.e., smaller items should be *inside* larger rather than the reverse. For example, you have `strong(p("Valid Example:"))` the better ordering would be `p(strong("Valid Example:))`
      - However, I would strongly consider using a heading tag here instead as this gives a better structure to your page.
    
    c. Please end all CSS with a semi-colon as a safety precaution. For example, `style = "text-align: center;"`
10. DRY Coding--look through your app and see if you have adhered to the DRY principle: Don't Repeat Yourself. If you find yourself repeating the same code multiple times, you need to step back and see if there is something different you can do. For example:
    - Would defining a [global] function be of better use? 
    - Would reorganizing a set of tabs inside a page be better? For example, each tab of the Explore page deals with Normality of residuals and you use essentially the same code. Would a better organization be tabbing through assumptions rather than models? (Hint: I think this would be worth exploring.)
    - Would removing the tabs from the Drag and Drop game and replacing that with a drop-down `selectInput` for the user to choose from in conjunction with the drag-and-drop be better? (Hint: YES!)
11. I think that we need to go back to the drawing board for the Multiple Choice Game. What other kinds of games might work with the idea of deciding whether or not an assumption is violated?
12. Move all code creating data frames and/or reading in data files to the section marked "Global Constants, Functions, and Data Sets". You'll still be able to call them by their names in the server function but this will provide us better performance.
13. Add `rstatix` and `lme4` to the library calls (otherwise these packages won't get loaded correctly).
14. Revamp the grading logic
  - What you currently have:
```
observeEvent(input$submitAnova,{
    output$markAnova <- renderUI({
      if ("Normality" %in% input$dropAnova &&
          "Homoscedasticity" %in% input$dropAnova &&
          "Independence of Observation" %in% input$dropAnova &&
          length(input$dropAnova) == 3){
        img(src = "check.png",width = 30)
      }
      else{
        img(src = "cross.png",width = 30)
      }
    })
  })
```
  - What would be better:
    - In global constants: `anovaAssumptions <- c("Normality", "Homoscedasticity", "Independence of Observations")`
```
observeEvent(input$submitAnova, {
  output$markAnova <- boastUtils::renderIcon(
      icon = ifelse(
          all(input$dropAnova %in% anovaAssumptions) && 
          length(input$dropAnova) == length(anovaAssumptions), 
          "correct",
          "incorrect
        )
     )
})
```

This is better as instead of looking for the correct answers in the user's response, you're looking for the user's response in the set of correct answers and you'd now ensure that the lengths would be the same. This would also give us some freedom to do a partial correct option if you nested an additional `ifelse` but were missing something (i.e., `length(input$dropAnova) < length(anovaAssumptions)`).

  - These changes can be carried over to the other parts of that game. (Remember, DRY.)
  - Additionally, using `boastUtils::renderIcon` can be used elsewhere as well

15. Please switch `icon("dashboard")` to `icon("tachometer-alt")` this is to try to prevent a future breakage for the icon.
16. Address Issue #2

I'm sure that I will have many more things for you to work on, but this list is already quite large. Tackle the items I've listed here and then we can continue on with reviewing.

lghszdnp committed:
Minor Change

lghszdnp committed:
Merge branch 'Demo' of https://github.com/EducationShinyAppTeam/Assumptions_of_ANOVA into Demo

lghszdnp committed:
Minor change

lghszdnp committed:
Fixed Some Problem

1. Fixed
2. Fixed
3. Fixed
4. Fixed
5. I think we can add a button of 'Next Page' in every tab we have (I will try to do this and show the new version on friday's meeting), or we can delete all the bsBotton because the student can already use the sidebar to choose the tabs.
6. I will ask my teammate to double check this.
7. Fixed
8. Fixed
9. I'm still working on the 80 characters limit. I have some trouble with the RStudio's guide but I think I will fix this before friday.
10. Fixed
11. Stilling working on it
12. I will consider about it and maybe we can discuss this during friday's meeting.
13. Fixed
14. Fixed
15. Still working on it as DRY Coding
16.  Fixed
17.  Still working on it

lghszdnp committed:
Wording Problems

Fixed some wording problems in this version

neilhatfield reviewed (changes requested):
Further Changes:

1. Remove the square brackets, [ ], from `APP_TITLE`.
2. Feedback/comments icon needs to be the to left of the Home button and you need to properly format the last bit: it should be `?appName=Assumptions_of_ANOVA`
3. You may delete lines 69-75. The `boastApp` call automatically links the boast.CSS file. Additionally, lines 74-75 don't actually do anyting.
4. You still have long lines of code which need to be resolved. In RStudio, enter into the Global Options found in the Tools menu. Click on Code in the left menu bar of the Options window that comes up. In the resulting window, there will be a tab for Display along the top. Click that. Check the box next to Show margin and ensure that Margin column is set to 80. Press OK. This will create a thin, grey line down the right-hand side of your coding window. Any time you have a line that goes beyond this line, you will need to break the line. The exception is a URL, which you should not break.
5. Instructions list
    + Change "go" in the first instruction to "GO!" to better match the button.
    + The second instruction is not an instruction, remove.
    + Add a comma after "section" in the third instruction (which will become the second instruction).
6. Place your name first in the Acknowledgements.
7. Prerequisites
    + New first box
        - Title: What is ANOVA?
        - Content: "In general, ANOVA refers to a family of statistical techniques that assess potential differences in a response (i.e., scale-level dependent variable) given one or more factors (i.e., nominal-level independent variables with 2+ categories). Each specific model in the ANOVA family has their own assumptions."
    + Second box
        - Title: Why is assumption testing important to ANOVA?" 
        - Content: ANOVA methods are primarily used for statistical inference. In order for the inference to be valid (regardless of whether we reject/fail to reject the null hypothesis), the assumptions which underpin that inference must be satisfied. If the assumptions aren't, then any p-values or confidence intervals we found don't mean what we want them to mean.
    + Third box
        - Title: What if the assumptions aren't met?
        - Content: By checking the assumptions before doing inference, we give ourselves the opportunity to shift gears. We can do this in a number of ways including but not limited to: using a different model, transforming the response, using a robust method, using randomization methods, and using nonparametric rank-based methods.
8. Please better format your code. For example, if you move first arguments to their own line, this will give subsequent arguments more room by making the alignment further to the left.
9. Be explicit; include all argument names.  See the coding example below for how to improve both 8 and 9.
10. I think that there is still too much repetition by having the tabs go through models. I feel that this layout detracts from the assumptions aspect.
11. The above comment notwithstanding, we need to address the layout of the Explore. 
    + I would like for you to place all instances of "Scenario" in `h3` (no `strong`, no `p`).
    + Get rid the `fluidRow` calls. They are not necessary for your app and you aren't using them correctly.
    + Convert both "Valid Example" and "Invalid Example" to `h4` (no `p`, no `tags$li`). You need more structure to your document, which `h4` provides and you aren't using `tags$li` (or `p`) correctly.
    + To help you out, below I've re-written the code for the ANOVA tab. Use this as your template for items 8, 9, and 11. 

```
tabsetPanel(
  id = "exploreModels",
  ##### Set up ANOVA page ----
  tabPanel(
    title = "Oneway ANOVA",
    br(), # This adds a bit of space between the tabs and content
    h3("Scenario"),
    p("An experiment was conducted to determine the relationship of
       honey output and the types of flowers,there were 3 types of
       flowers, 3 beehives were randomly assigned to each type of
       flowers with 9 beehives in total."),
    selectInput(
      inputId = "anovaSelect",
      label = "Select the assumption you want to explore",
      choices = list(
        "Normality of Residuals" = "normality",
        "Homoscedasticity" = "homoscedasticity",
        "Independence of Observation" = "independence"
      ),
      width = NULL,
    ),
    h4("Valid Example"),
    textOutput("anovaTextValid"),
    plotOutput("anovaImageValid"),
    tags$script(HTML(
      "$(document).ready(function() {
      document.getElementById('anovaImageValid').setAttribute('aria-label',
      `This plot output is depend on the user's choice. For normality,
      this is a plot of norm qunatile with all the points lay in the
      97% confidence envelope. For homoscedasticity, this is a plot
      of response versus explanatory, all the points lay in a random
      position. For independence, this is a plot of response versus
      index, all the points lay in a random position.`)
      })"
    )),
    h4("Invalid Example"),
    textOutput("anovaTextInValid"),
    plotOutput("anovaImageInValid"),
    tags$script(HTML(
      "$(document).ready(function() {
      document.getElementById('anovaImageInvalid').setAttribute('aria-label',
      `This plot output is depend on the user's choice. For normality,
      this is a plot of norm qunatile with 2 points lay in the 97%
      confidence envelope. For homoscedasticity, this is a plot of
      response versus explanatory, all the points lay in a pattern.
      For independence, this is a plot of response versus index, all
      the points lay in a pattern.`)
      })"
    ))
  ),
  tabPanel(
    ... # This is where you start the next tab
```

12. Use the `width` argument of `selectInput` to control how wide the drop down menu is for each set of assumptions. If the menu is wide enough be default, use `width = NULL` (as I did for the ANOVA tab). However, if you have options that are wrapping, play around with using `"##%"`. (You will need the double quotation marks and the percent sign.) Explore what values to use instead of ## that will make the menu look better. Note: this is a percentage of the space (line of the page) that the menu is located in.
13. There are still spelling and grammar errors. For example,
    + The text that you're rendering in `textOutput` should be complete sentences (including end of sentence punctuation). 
    + There are words misspelled in the aria-labels (e.g., "qunatile").
    + "Points" does not need to be capitalized if it isn't the first word of a sentence.
    + Note: I did not fix these in my above code example.
14. The aria-labels need to be fixed. You can't just stick all of the various plot's descriptions into a static aria-label and call it a day. That makes the label rather confusing for a person using it. You will need to make the aria text more dynamic. Leah has some examples of this in both the Homogeneous and Homogeneous Poisson Process apps. (Note: I did not fix this in my above code example.)
15. Arguments to add to all `car::qqplot` calls:
    + `pch =19`
    + `cex = 1.5`
    + `id = FALSE`
    + If you can figure out how to increase the font size of the labels, that would be great.
16. Arguments to add/change to all `stripchart` calls:
    + `pch =19`
    + `cex = 1.5`
    + You'll need to play around with the values for `cex.lab` (for axis labels) and `cex.axis` (for values on the axes) to find the optimum font sizes.
17. Arguments to add/change to all `plot` calls:
    + `pch =19`
    + `cex = 1.5`
    + You'll need to play around with the values for `cex.lab` (for axis labels) and `cex.axis` (for values on the axes) to find the optimum font sizes.
    + Use `xlab = "Row index"` for any of the plots for independence of observations for additional clarity
18. Linear relationship between covariate and response
    + Remove the keyboarding type (color) from the plot.
    + Add the appropriate trend line (e.g., a linear smoother)
    + For the invalid example, you need to make this look like an even bigger violation. For example, something truly quadratic.
    + I think what is happening is that you're conflating the equality of slope parameters with the linear relationship between the covariate and response.
        - For the linear relationship you only need to examine the response and the covariate (ignore the factor) and see if there is a linear relationship or not.
        - For the equality of slop parameters, this is where you bring in the factor and are wanting to see if there are parallel lines.
    + Adjust the font sizes of these plots see [Style Guide-Plot Font](https://educationshinyappteam.github.io/Style_Guide/designStyle.html#text-in-r-plots)
19. Equality of Slope parameters: use both color and shape convey the factor (this is more accessible than just using color).
20. Why do you suddenly start using blue colored points for Random Effects Repeated Measures (Homoscedasticity)? Please be consistent throughout all charts in your usage of color. (Preference is black, filled in dots when you aren't needing to highlight a third aspect.)
21. For Repeated Measures, please replace 
    + "HVBC" with "Happy Valley Brewing Company
    + "Otto's" with "Otto's Pub and Brewery"
    + "Robin Hood" with "Robin Hood Brewing Co."
22. Convert to using `renderIcon` for all grading in the games.
23. If you are going to keep the Multiple Choice game present, then you need to shuffle the location of the correct answer. I know that we haven't been able to come up with a programmed way to do this and I'm not asking you to solve that problem now. What I am asking you to do is make it so that option C isn't always the correct answer. The easiest solution here is to use `sample(x = letters[1:3], size = 7, replace = TRUE)` to give you a [pseudo-] randomized ordering and then you program the plots in the appropriate way. This will keep the fixed ordering issue we've yet to solve but will at least make it so that C isn't always the correct answer.
    + Otherwise, we should remove the Multiple Choice game from the app.
24. Please place all references in the AmStat style (See [Style Guide-References](https://educationshinyappteam.github.io/Style_Guide/documentation.html#references) for examples and links to the AmStat website) as well as putting them in alphabetical order by first author's last name.
25. Make sure to update the "Last Update" line on the overview.

Be on the look out for an additional email from me since you want to demo this app at the webinar.

lghszdnp committed:
Update

I fixed most problems you mentioned in the feedback. Only the aria-labels and locations of multiple choice game is left. For the multiple choice game, I have figured out how to shuffle their location, but I'm still working on build an interaction between the plot and the grade box, hope I will figure it out before our next meeting. I also something new in the drag and drop game: there will be a hint about what kind of mistake you made in the game.

lghszdnp committed:
Description and README

lghszdnp committed:
Update

 I set the answers of multiple-choice games to other choices rather than 'c'. but it won't change when you refresh the page.

Update some plots in the explore page to make it more clarified

neilhatfield committed:
Update app.R

Clean up and reorganization of code

neilhatfield reviewed (approved):
I'm approving my upload of the code.
